docs: Add Bubble integration docs for workflows and notifications (#1721)
apify/apify-docs: sources/platform/integrations/workflows-and-notifications/bubble.md

---
title: Bubble integration
description: Learn how to integrate your Apify Actors with Bubble for automated workflows and notifications.
sidebar_label: Bubble
sidebar_position: 7

> slug: /integrations/bubble

Author: done ✅

---

**Learn how to integrate your Apify Actors with Bubble for automated workflows and notifications.**

---

[Bubble](https://bubble.io/) is a no-code platform that allows you to build web applications without writing code. With the Apify integration for Bubble, you can easily connect your Apify Actors to your Bubble applications to automate workflows and display scraped data.

## Get started

To use the Apify integration for Bubble, you will need:

- An [Apify account](https://console.apify.com/)
- A [Bubble account](https://bubble.io/)
- A Bubble application where you want to use the integration

## Step 1: Install the Apify plugin for Bubble

To integrate Apify with your Bubble application, you first need to install the Apify plugin from the Bubble plugin marketplace.

> ![Apify Plugin Download](../images/bubble/plugin_install_preview.png)

Author: done ✅

> 1. Go to your Bubble application dashboard and navigate to the **Plugins** tab.

Reviewer: You mean the 1. in each point in ordered list? This is correct way of doing it, docusaurus will then on build numbers proper.

> 1. Go to your Bubble application dashboard and navigate to the **Plugins** tab.

Reviewer: Just FYI the reasons for it is mostly maintenance, it's easier to make it all `1.` and let docusaurus handle numbering than number them yourself. Furthermore when you need to adjust the steps (remove or add or whatever) you don't have to think about numbering them manually.

1. Click the **Add plugins** button.
1. Search for "Apify" in the plugin marketplace.
1. Click on the Apify plugin and then click **Install**.

## Step 2: Configure the Apify plugin

After installing the plugin, you'll need to provide your API token when setting up Apify actions:

1. You can find your API token in the Apify Console under **Settings > API & Integrations**.

> 2. For security, store your API token in your User data type with Privacy rules applied rather than entering it directly in each API call.

Author: I tried to make it clearer. I mentioned how to set it up and how to use it. Tomorrow i will also link some bubble docs, so the user has enough info to follow the guide.

## Using the integration

Once the plugin is configured, you can start building automated workflows.

### Trigger Actor runs from Bubble events

> Create workflows that run Apify Actors in response to events in your Bubble app, such as button clicks or form submissions. In the **Workflow** tab, set up a new workflow with your chosen trigger event. Add a new action and select **Run Actor** or **Run Actor Task** from the Apify plugin. Configure the Actor or Task, including any input data, and include follow-up actions to process the returned data.

Author: Tried to make it clearer, let me know if you would add/remove anything from the current version. I will review it again tomorrow to ensure everything makes sense.

2. Configure the specific Actor or task you want to display
3. Use "Fetch Data From Dataset Json As Data" to show actual scraped results from your Actor runs

![Using Apify data calls](../images/bubble/data_preview.png)

### Automate data collection workflows

Set up recurring data collection that updates your Bubble app automatically:

1. Create a scheduled workflow or API workflow
2. Use the "Run Actor Task" action to collect fresh data
3. Process and store the results in your Bubble database
4. Update your app's display elements with new data

![Using Apify Actions calls](../images/bubble/automate_data_workflow.png)

## Step 4: Example use cases

### E-commerce price monitoring

1. Set up a workflow that triggers daily or weekly
2. Use the "Run Actor" action to scrape competitor prices
3. Display the results in your Bubble app's dashboard
4. Set up notifications when prices change significantly

### Lead generation automation

1. Create a form submission workflow
2. Use scraped data to enrich lead information automatically
3. Store enriched data in your Bubble database
4. Trigger follow-up actions based on the enriched data

> ### Content aggregation
> 
> 1. Schedule regular Actor runs to collect content from multiple sources
> 2. Use "Fetch Data From Dataset Json As Data" to display aggregated content
> 3. Create filtered views based on content categories or dates

Reviewer: ```suggestion
### Content aggregation

Configure regular Actor runs to gather articles or posts from multiple sources. Use the **Get Dataset Items** data call to display the content in your app, and enable user filters by category, date, or source.
```

   - ![Create a datatype with new field](../images/bubble/step3_scraping.png)
   - Create the backend workflow (`webhook`) that Bubble exposes at `/api/1.1/wf/webhook`. The workflow name defines the API route.
   - ![Create a backend webhook](../images/bubble/step4_scraping.png)
   - In that workflow, for each received webhook call, create a new thing in `ScrapingResults` and set `result` to the dataset ID from the request body. This stores one `datasetId` per call for later processing.
   - ![Add new result](../images/bubble/step5_scraping.png)
1. Pick up the results asynchronously
   - In a (periodic) backend workflow, search `ScrapingResults` for pending entries (or for the expected `datasetId`).
   - If found, read its `result` (the `datasetId`), fetch items via the appropriate action (for example, **Fetch Data From Dataset JSON As Action**), update the UI or save to your DB, and then delete that `ScrapingResults` entry to avoid reprocessing.
   - If not found yet, do nothing and check again later.
   - ![Do every time](../images/bubble/step6_scraping.png)

This approach avoids Bubble timeouts, keeps the UI responsive, and scales to larger scrapes.

## Example use cases

> - _E-commerce price monitoring_  - Schedule a daily workflow to run a price-scraping Actor on competitor sites. Store the results in your Bubble database, display them in a dashboard, and set up alerts for significant price changes.

Author: Good catch, thanks

- _Lead generation automation_ - Trigger a workflow on form submission to run an Actor that enriches lead data, such as pulling company details from a domain. Save the enriched information to your database and automate follow-up actions like email campaigns.
- _Content aggregation_ - Configure regular Actor runs to gather articles or posts from multiple sources.

## Available Apify actions and data sources

The Apify plugin provides two main types of operations:

> **Data calls** (data sources):
> - Fetch Data From Dataset Text As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
> - Fetch Data From Dataset File As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
> - Fetch Data From Dataset Json As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
> - List User Datasets [doc](https://docs.apify.com/api/v2/datasets-get)
> - List Key-Value Stores [doc](https://docs.apify.com/api/v2/key-value-stores-get)
> - List Store Keys [doc](https://docs.apify.com/api/v2/key-value-store-keys-get)
> - Get Record As File [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
> - Get Record As Image File [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
> - List Store Actors [doc](https://docs.apify.com/api/v2/store-get)
> - List User Actors [doc](https://docs.apify.com/api/v2/acts-get)
> - List User Tasks [doc](https://docs.apify.com/api/v2/actor-tasks-get)
> - Scrape Single URL As File As Data [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
> - Scrape Single URL As Json As Data [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
> - List Actor Runs [doc](https://docs.apify.com/api/v2/act-runs-get)
> - List Specific Actor Runs [doc](https://docs.apify.com/api/v2/act-runs-get)
> - List Webhooks [doc](https://docs.apify.com/api/v2/webhooks-get)
> 
> **Actions** (workflow steps):
> - Fetch Data From Dataset Text As Action [doc](https://docs.apify.com/api/v2/dataset-items-get)
> - Fetch Data From Dataset File As Action [doc](https://docs.apify.com/api/v2/dataset-items-get)
> - Get Record As Text [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
> - Scrape Single URL As File As Action [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
> - Scrape Single URL As Json As Action [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
> - Run Actor [doc](https://docs.apify.com/api/v2/act-runs-post)
> - Run Actor Task [doc](https://docs.apify.com/api/v2/actor-task-runs-post)
> - Create Webhook [doc](https://docs.apify.com/api/v2/webhooks-post)
> - Delete Webhook [doc](https://docs.apify.com/api/v2/webhook-delete)

Reviewer: why add separate [doc] next the Data call/Action instead of making it link to specific endpoints?

The you could frontload the admonition below before the list like this

```suggestion
tip::: Check out the documentation

Each API call links to the Apify documentation. To learn more about any plugin action or data call, go to the **Plugins** page in your app, select the Apify plugin, and use the documentation links in the field descriptions.

:::

**Data calls** (data sources):
- Fetch Data From Dataset Text As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
- Fetch Data From Dataset File As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
- Fetch Data From Dataset Json As Data [doc](https://docs.apify.com/api/v2/dataset-items-get)
- List User Datasets [doc](https://docs.apify.com/api/v2/datasets-get)
- List Key-Value Stores [doc](https://docs.apify.com/api/v2/key-value-stores-get)
- List Store Keys [doc](https://docs.apify.com/api/v2/key-value-store-keys-get)
- Get Record As File [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
- Get Record As Image File [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
- List Store Actors [doc](https://docs.apify.com/api/v2/store-get)
- List User Actors [doc](https://docs.apify.com/api/v2/acts-get)
- List User Tasks [doc](https://docs.apify.com/api/v2/actor-tasks-get)
- Scrape Single URL As File As Data [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
- Scrape Single URL As Json As Data [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
- List Actor Runs [doc](https://docs.apify.com/api/v2/act-runs-get)
- List Specific Actor Runs [doc](https://docs.apify.com/api/v2/act-runs-get)
- List Webhooks [doc](https://docs.apify.com/api/v2/webhooks-get)

**Actions** (workflow steps):
- Fetch Data From Dataset Text As Action [doc](https://docs.apify.com/api/v2/dataset-items-get)
- Fetch Data From Dataset File As Action [doc](https://docs.apify.com/api/v2/dataset-items-get)
- Get Record As Text [doc](https://docs.apify.com/api/v2/key-value-store-record-get)
- Scrape Single URL As File As Action [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
- Scrape Single URL As Json As Action [doc](https://docs.apify.com/api/v2/act-run-sync-get-dataset-items-post)
- Run Actor [doc](https://docs.apify.com/api/v2/act-runs-post)
- Run Actor Task [doc](https://docs.apify.com/api/v2/actor-task-runs-post)
- Create Webhook [doc](https://docs.apify.com/api/v2/webhooks-post)
- Delete Webhook [doc](https://docs.apify.com/api/v2/webhook-delete)
```

> :::tip
> 
> ### Check out the documentation
> Each API call links to the Apify documentation. To learn more about any plugin action or data call, go to the **Plugins** page in your app, select the Apify plugin, and use the documentation links in the field descriptions.
> 
> ### Use the latest version of the plugin
> To stay up to date with new features, make sure you're using the latest version of the plugin. You can check this on the **Plugins** page by selecting the Apify plugin and choosing the latest version from the drop-down menu. You'll also see a brief note describing what's changed in that version.
> 
> :::

Reviewer: This cannot be like that. I would suggest frontloading the first tip above the list and then leave only second tip here

```suggestion
:::tip

### Use the latest version of the plugin

To stay up to date with new features, make sure you're using the latest version of the plugin. You can check this on the **Plugins** page by selecting the Apify plugin and choosing the latest version from the drop-down menu. You'll also see a brief note describing what's changed in that version.

:::
```

![Apify Plugin's data calls](../images/bubble/data_calls_preview.png)

::::tip
Each Apify plugin action and data call input in Bubble includes inline documentation describing what the parameter is for and the expected format. If you're unsure, check the field's help text in the Bubble editor.
::::

---

### Dynamic values in inputs and data calls

Dynamic values are available across Apify plugin fields. Use Bubble's **Insert dynamic data** to bind values from your app.

- For instance you can source values from:
  - **Page/UI elements**: inputs, dropdowns, multi-selects, radio buttons, checkboxes
  - **Database Things and fields**
  - **Current User**
  - **Previous workflow steps** (e.g., Step 1's Run Actor result's `defaultDatasetId` or `runId`)
  - **Get Data from an External API**: data calls

Examples
1) Use a page input in an Action's JSON field (Input overrides):

```json
{
  "url": "Input URL's value"
}
```
![Select Insert dynamic value](../images/bubble/insert_dynamic_data.png)

2) Use a previous step's result in JSON:

```json
{
  "datasetId": "Step 1's Run Actor's defaultDatasetId"
}

> ```

Reviewer: huh this looks weird in suggestion mode :D

![Repeating group example](../images/bubble/repeating_group.png)

2) List user datasets

- This example lists the current user's datasets and displays them in a repeating group.
- Add a **Repeating group** to the page.
  - Step 1 — Add data to a variable: create a custom state (for example, on the page) that will hold the list of datasets, and set it to the plugin's **List User Datasets** data call.
![Step 1 — Set variable with user's datasets](../images/bubble/user_dataset_repeating_group_set.png)
  - Step 2 — Set the type: in the repeating group's settings, set **Type of content** to match the dataset object your variable returns.
![Step 2 — Repeating group type of content](../images/bubble/user_dataset_repeating_group.png)
  - Step 3 — Bind the variable: set the repeating group's **Data source** to the variable from Step 1.
![Step 3 — Repeating group data source](../images/bubble/user_dataset_repeating_group_source.png)
- Inside the repeating group cell, bind dataset fields (for example, `Current cell's item name`, `id`, `createdAt`).

> ![Step 4 — Repeating group data cell](../images/bubble/user_dataset_repeating_group_cell.png)

Reviewer: Why suddenly 1) instead of 1.?

also these should be H4's

![Create a datatype with new field](../images/bubble/step3_scraping.png)
   - Create the backend workflow (`webhook`) that Bubble exposes at `/api/1.1/wf/webhook`. The workflow name defines the API route.
![Create a backend webhook](../images/bubble/step4_scraping.png)
   - In that workflow, for each received webhook call, create a new thing in `ScrapingResults` and set `result` to the dataset ID from the request body. This stores one `datasetId` per call for later processing.
![Add new result](../images/bubble/step5_scraping.png)

4. Pick up the results asynchronously
   - In a (periodic) backend workflow, search `ScrapingResults` for pending entries (or for the expected `datasetId`).
   - If found, read its `result` (the `datasetId`), fetch items via the appropriate action (for example, **Fetch Data From Dataset JSON As Action**), update the UI or save to your DB, and then delete that `ScrapingResults` entry to avoid reprocessing.
   - If not found yet, do nothing and check again later.
![Do every time](../images/bubble/step6_scraping.png)

This approach avoids Bubble timeouts, keeps the UI responsive, and scales to larger scrapes.

> ---

Reviewer: What is the purpose of this line?

## Example use cases

> ### E-commerce price monitoring
> 
> Schedule a daily workflow to run a price-scraping Actor on competitor sites. Store the results in your Bubble database, display them in a dashboard, and set up alerts for significant price changes.
> 
> ### Lead generation automation
> 
> Trigger a workflow on form submission to run an Actor that enriches lead data, such as pulling company details from a domain. Save the enriched information to your database and automate follow-up actions like email campaigns.
> 
> ### Content aggregation
> 
> Configure regular Actor runs to gather articles or posts from multiple sources. Use the **Fetch Data From Dataset JSON As Data** call to display the content in your app, and enable user filters by category, date, or source.

Reviewer: This feels dry? I would argue that with so little description these might have been an unordered list? Maybe something like this

```suggestion
- _E-commerce price monitoring_  - Schedule a daily workflow to run a price-scraping Actor on competitor sites. Store the results in your Bubble database, display them in a dashboard, and set up alerts for significant price changes.

- _Lead generation automation_ - Trigger a workflow on form submission to run an Actor that enriches lead data, such as pulling company details from a domain. Save the enriched information to your database and automate follow-up actions like email campaigns.

- _Content aggregation_ - Configure regular Actor runs to gather articles or posts from multiple sources. Use the **Fetch Data From Dataset JSON As Data** call to display the content in your app, and enable user filters by category, date, or source.
```

- List Key-Value Stores
- List Store Keys
- Get Record As File
- Get Record As Image File
- List Store Actors
- List User Actors
- List User Tasks
- Scrape Single URL As File As Data
- Scrape Single URL As Json As Data
- List Actor Runs
- List Specific Actor Runs
- List Webhooks

**Actions** (workflow steps):
- Fetch Data From Dataset Text As Action
- Fetch Data From Dataset File As Action
- Get Record As Text
- Scrape Single URL As File As Action
- Scrape Single URL As Json As Action
- Run Actor
- Run Actor Task
- Create Webhook
- Delete Webhook

---

## Troubleshooting
<!-- markdownlint-disable MD001 -->
#### Authentication errors

Ensure your API token is correctly set in the action (preferably as `Current User's apify_api_token`) and that it has the necessary permissions.

#### Missing Actors or tasks 

If an Actor or task doesn't appear in a dropdown, make sure you have run it at least once from the Apify Console.

#### Timeout errors

Bubble workflows have a timeout limit. For long-running Actors, consider setting **Wait until finished** = no, then process results in a separate backend workflow triggered by a webhook or scheduled event.

#### Data format issues

> Check that your JSON input is valid when providing **Input overrides** and that dynamic expressions resolve to valid JSON values. Verify the structure of the dataset output when displaying it in your app.

Reviewer: Why not H3?

Author: Yeah, I was reediting it, and all the weird formatting parts that you mentioned were the result of it...  I've addressed this in the latest commit. Let me know if that version fixed all that you mentioned. Thanks.
